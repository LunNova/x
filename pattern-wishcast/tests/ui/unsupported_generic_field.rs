// SPDX-FileCopyrightText: 2025 LunNova
//
// SPDX-License-Identifier: MIT

//! Test that unsupported generic fields containing Self produce helpful errors

use pattern_wishcast::pattern_wishcast;
use std::collections::HashMap;

pattern_wishcast! {
	enum StuckEvaluation = {
		Var { id: usize },
	};

	enum Value is <P: PatternFields> = StuckEvaluation | {
		HostValue { value: String },
		// This should produce a compile error:
		BadField { data: HashMap<String, Self> },
	};

	type FlexValue = Value is _;
	type StrictValue = Value is HostValue(_) | BadField(_);

	#[derive(SubtypingRelation(upcast=to_flex, downcast=try_to_strict))]
	impl StrictValue : FlexValue;
}

fn main() {}
