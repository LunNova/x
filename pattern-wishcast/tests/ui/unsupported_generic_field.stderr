error: Unsupported field type: HashMap containing Value types. Only Vec<T>, Box<T>, and Option<T> are supported for generic containers. Field: data
  --> tests/ui/unsupported_generic_field.rs:12:1
   |
12 | / pattern_wishcast! {
13 | |     enum StuckEvaluation = {
14 | |         Var { id: usize },
15 | |     };
...  |
27 | |     impl StrictValue : FlexValue;
28 | | }
   | |_^
   |
   = note: this error originates in the macro `pattern_wishcast` (in Nightly builds, run with -Z macro-backtrace for more info)
